src/components/implementation/torrent/ramfs/ramfs.c

```diff
@@ -186,6 +244,139 @@ twrite(spdid_t spdid, td_t td, int cbid, int sz)
 	return ret;
 }
 
+int 
+twritep(spdid_t spdid, td_t td, int cb, int start, unsigned int sz)
+{
+	printc(">>entering twritep\n");
+	int ret = -1, left;
+	struct torrent *t;
+	struct fsobj *fso;
+	__file_data_t *file_data; 
+
+	printc(">>>starting basic checks\n");
+	if (!sz) return -EINVAL;
+
+	if (tor_isnull(td)) return -EINVAL;
+
+	LOCK();
+
+	t = tor_lookup(td);
+	if (!t) ERR_THROW(-EINVAL, done);
+	assert(t->data);
+	if (!(t->flags & TOR_WRITE)) ERR_THROW(-EACCES, done);
+
+	fso = t->data;
+	assert(fso->size <= fso->allocated);
+	assert(t->offset <= fso->size);
+	__file_data_t *current = (__file_data_t*) fso->data; // start of file
+	__file_data_t *prev = NULL;
+
+	// find where we want to write.
+	printc(">>>figuring out where we want to write\n");
+	u32_t total_offset = 0;
+	while (total_offset < t->offset && current != NULL)
+	{
+		printc(">>>>still thinking about it\n");
+		total_offset += current->len;
+		prev = current;
+		current = current->next;
+		// crap. Current will alwyas be NULL
+	}
+
+	// need to calculate proper start
+
+	if (current == NULL)
+	{
+		// initial file creation
+
+		printc(">>>case 0\n");
+		// adding to start and can't override
+		__file_data_t *new_node = malloc(sizeof(__file_data_t));
+		new_node->cbuf_id = cb;
+		new_node->cbuf_len = sz; // it reeally doesn't.
+		new_node->start = start;
+		new_node->len = sz;
+		new_node->next = NULL;
+
+		fso->data = new_node; // change start of file
+
+		// set offset
+		t->offset += sz;
+		fso->size = t->offset;
+
+		printc(">>>exiting case 0\n");
+	}
+	else if (t->offset == 0 && sz < current->len)
+	{
+		printc(">>>case 1\n");
+		// adding to start and can't override
+		__file_data_t *new_node = malloc(sizeof(__file_data_t));
+		new_node->cbuf_id = cb;
+		new_node->cbuf_len = sz; // it reeally doesn't.
+		new_node->start = start;
+		new_node->len = sz;
+		new_node->next = current->next;
+
+		// need to keep some of current
+		current->len -= sz;
+
+		fso->data = new_node; // change start of file
+
+		// set offset
+		t->offset += sz;
+		fso->size = t->offset;
+	}
```

Comment:
The difference between the above two blocks of code are minimal. You could make them one if () { } block easily and reduce LOC.